### ISSUE TITLE
Packed Ray Operations

### ISSUE THREAD

**Issue by tovacinni** [NVIDIAGameWorks/kaolin#462]
When dealing with rays on top of a data structure like SPC (or any BVH-like structure in general), each ray will have a variable number of intersections. The 'trivial' way to deal with this is to have a padded dense tensor where one dimension is the max sequence length, but in cases where the sequence lengths heavily vary (as is the case in ray BVH intersections) this is wasteful in both compute and memory.

An alternative methodology (adopted in other places within Kaolin) is to use `packed` tensors, which are variable length sequences flattened into a tensor. This also implies we need custom operations and kernels to deal with these types of data structures. (It's important to also be aware that in PyTorch there are `torch.index_*` operations which would be good to compare against in terms of perf for some ops). 

This includes the following:

- cumulative sum
- cumulative product
- sum
- product
- exponential integration
- ray sampling + distance calculation

and maybe some others. Currently many of these are implemented in an internal draft MR but are still missing backprop on some operations like `cumulative product`.

**Comment by tovacinni** [comment]
A note on backprop for `cumulative product`: 

In general, `cumprod` maps a sequence to a sequence so the Jacobian will be `O(n^2)`. One can speed up the Jacobian-vector product by computing the `cumprod` once and using `cumsum` and a division (and hence reusing the product) for a `O(n)` algorithm, but the division can result in `NaN`s if the input has any zeros. Different frameworks handle this differently, which I'll highlight:

In TensorFlow, they completely ignore the divide by zero issue and treat `0/0 = 0` , which results in a `O(n)` but an incorrect gradient. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L1995

In PyTorch <1.9, they adopt the `O(n^2)` approach in presence of zeros, which is correct but slower. In PyTorch 1.9 (released June 2021) they merged an update which uses an `O(n)` algorithm based on `cumsum`, with some additional case analysis and masks to handle the 0 case. https://github.com/pytorch/pytorch/pull/53711
A lot of NeRF reimplementations from TensorFlow to PyTorch claim that there is a 1to1 numerical correspondence which might be true for the forward pass but I don't think is true for the backward pass. In general, PyTorch 1.9 seems like it will give both a correct and fast implementation, but I don't think a benchmark exists anywhere.
